charger/mcc.go

```diff
@@ -0,0 +1,352 @@
+		uri:        strings.TrimRight(uri, "/"),
+		password:   password,
+	}
+
+	return mcc
+}
+
+// construct the URL for a given apiFunction
+func (mcc *MobileConnect) apiURL(api apiFunction) string {
+	return fmt.Sprintf("%s/%s", mcc.uri, api)
+}
+
+// proces the http request to fetch the auth token for a login or refresh request
+func (mcc *MobileConnect) fetchToken(request *http.Request) error {
+	var tr MCCTokenResponse
+	b, err := mcc.RequestJSON(request, &tr)
+	if err == nil {
+		if len(tr.Token) == 0 && len(b) > 0 {
+			var error MCCErrorResponse
+
+			if err := json.Unmarshal(b, &error); err != nil {
+				return err
+			}
+
+			return fmt.Errorf("response: %s", error.Error)
+		}
+
+		mcc.token = tr.Token
+		// According to the Web Interface, the token is valid for 10 minutes
+		mcc.tokenValid = time.Now().Add(10 * time.Minute)
+
+		// the web interface updates the token every 2 minutes, so lets do the same here
+		mcc.tokenRefresh = time.Now().Add(2 * time.Minute)
+	}
+
+	return err
+}
+
+// login as the home user with the given password
+func (mcc *MobileConnect) login(password string) error {
+	uri := fmt.Sprintf("%s/%s", mcc.uri, apiLogin)
+
+	data := url.Values{
+		"user": []string{"user"},
+		"pass": []string{mcc.password},
+	}
+
+	req, err := http.NewRequest(http.MethodPost, uri, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	return mcc.fetchToken(req)
+}
+
+// refresh the auth token with a new one
+func (mcc *MobileConnect) refresh() error {
+	uri := fmt.Sprintf("%s/%s", mcc.uri, apiRefresh)
+
+	req, err := http.NewRequest(http.MethodGet, uri, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", mcc.token))
+
+	return mcc.fetchToken(req)
+}
+
+// creates a http request that contains the auth token
+func (mcc *MobileConnect) request(method, uri string) (*http.Request, error) {
+	// do we need to login?
+	if mcc.token == "" || time.Since(mcc.tokenValid) > 0 {
+		if err := mcc.login(mcc.password); err != nil {
+			return nil, err
+		}
+	}
+
+	// is it time to refresh the token?
+	if time.Since(mcc.tokenRefresh) > 0 {
+		if err := mcc.refresh(); err != nil {
+			return nil, err
+		}
+	}
+
+	// now lets process the request with the fetched token
+	req, err := http.NewRequest(method, uri, nil)
+	if err != nil {
+		return req, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", mcc.token))
+
+	return req, nil
+}
+
+// use http PUT to set a value on the URI, the value should be URL encoded in the URI parameter
+func (mcc *MobileConnect) putValue(uri string) ([]byte, error) {
+	req, err := mcc.request(http.MethodPut, uri)
+	if err != nil {
+		return nil, err
+	}
+
+	return mcc.Request(req)
+}
+
+// use http GET to fetch a non structured value from an URI and stores it in result
+func (mcc *MobileConnect) getValue(uri string) ([]byte, error) {
```

Comment:
Dito?

Reply:
Die wird in `Status()` und `Enabled()` benutzt